### DISCUSSION TITLE
How to use `command.lift`?

### DISCUSSION THREAD

**Discussion by eduludi** [ueberdosis/tiptap#1836] [General Questions] [answered]
We have an _inline_ Node for links. A link as mark was not enough, since we need to have a way to edit some custom aspects of the link.

After some try and error, we were able to add a link using the `insertContent` command, which looks hacky. 

The problem we face now is how to remove the link. We tried with both `lift` and `toggleWrap` commands, but they don't respond: the link wrapper is still there and we don't get any error or hint on what can be wrong. 

Here is an extract of what we currently have:

```js
export default Node.create({
  name: 'link',
  group: 'inline',
  inline: true,
  content: 'text*',

  addAttributes() {
    return {
      href: {
        default: '#',
      },
      // ... other attrs
    }
  },

  addCommands() {
    return {
      setLink: () => ({ commands, state }) => {
        const isActive = isNodeActive(state, 'link')
        if (isActive) {
          return commands.lift('link') // <== HERE
        } else {
          return commands.insertContent({
            type: 'link',
            attrs: {
              href: 'https://',
            },
            content: state.selection.content().toJSON().content[0].content,
          })
        }
      },
      removeLink: () => ({ commands }) => {
        return commands.lift('link') // <== AND HERE
      },
    }
  },
  // ...
}
```

PS: Really cool library, we are really excited as a team as it looks really promising and it will allow us to move from our existing solution (based on Slate.js) that has a lot of bugs.

**Comment by hanspagel** [comment, marked as answer]
I’d recommend to go with the mark-based approach. What "aspects" do you need to change? I bet you could do that with attributes.

https://www.tiptap.dev/guide/custom-extensions#attributes
https://www.tiptap.dev/api/commands/update-attributes
https://www.tiptap.dev/api/commands/extend-mark-range

**Reply by eduludi** [reply]
Thanks for your answer and recommendation. We will evaluate if links as `mark` will work for us.

Some benefits we see with the inline node approach is that we can render custom elements as we need. We will have more flexibility to achieve all kind of inline elements, as we do with Slate. And this is not only for links.

For example, this is what we have so far with a link as _inline node_ in `tiptap`:

<img width="369" alt="RichText Links with inline nodes" src="https://user-images.githubusercontent.com/76132/131983346-f9921e7f-121d-4ca5-a979-5cf7b75a3bb5.png">

> Note the external link icon on the left and the custom popover

With markers we feel we are restricted to do only some styling that we may modify through attributes, but that's it. There are other alternatives we are not aware of? 

But for now the issue is still that we cannot use `.lift()` or `.toggleWrap()` commands. Any ideas on why this may not be working?

Again, thanks for your answers and time :)

**Reply by hanspagel** [reply]
Got it! Did you try adding this popup with the bubble menu? https://www.tiptap.dev/guide/menus#bubble-menu

BTW your interface design is :chefkiss:

**Reply by eduludi** [reply]
Yes. We are already using the bubble menu for inline format:

<img width="402" alt="Screen Shot 2021-09-06 at 11 23 49" src="https://user-images.githubusercontent.com/76132/132193662-d4a10d4f-1522-4bb4-a5cf-3caf583dabb5.png">

We want to keep this bubble menu as simple as possible, and that's why we think adding a second popup for editing the link is a better option. We want to evaluate your suggestion, but we want first to rule out the inline node option. As you can see, we are really close to make it work, and that's why we opened this discussion to see if what we want to achieve is feasible. 

Again, thanks for your time and for this awesome editor called **tiptap**

**Reply by philippkuehn** [reply]
Hey, I think [`lift()`](https://prosemirror.net/docs/ref/#commands.lift) only works for block nodes – not inline nodes. For example for unwrapping a paragraph from a blockquote. There is no built-in command for what you are searching for yet, I think. You have to write a custom command for it: loop over nodes within your selection, find your `link` node, get its text content and replace the link node with a text node.

**Reply by eduludi** [reply]
Thanks @philippkuehn. That makes sense now.
